### PR TITLE
Add state safe party type

### PR DESCRIPTION
**Pull request by rnicoll**
Part of the identity design is that parties are represented on the ledger by their primary key only, so no identity details are revealed as part of the state. Party details are then resolved as needed, using local identity service records. This does introduce some new ways for things to fail, however enforces privacy.

### CODE REVIEW

**Comment by rnicoll** [comment]
This one's huge and invasive, so I'll add more reviewers later, but for now want to just get initial feedback. @exFalso as lead reviewer

**Comment by rick-r3** [comment]
I'm wondering if it's better to do `Party` as a sealed class with an `AnonymisedParty` and `IdentifiableParty` implementations?  Or otherwise have them interchangeable.  We'd then always store as the former (tx store could convert them), and run `verify()` as the former but resolve to the latter automagically for flows etc when needed, as our key store allows.  Possibly auto-harvest the identities and put in the key store when we are sent them.  I'm sure it will be useful to decorate the contract states with resolved identities.

Caveat: what does sealed class look like to Java?

What do others think?

**Comment by exFalso** [comment]
I only skimmed through this, will look at it in more detail on Monday.

Agree with @rick-r3, a sealed class would capture the intent in a clearer way. Afaik in Java a sealed class `sealed class A { class B : A() }` simply appears as a nested class `A.B` that subclasses `A`.

I **strongly** disagree with the mutable behaviour of `StateParty`. Iirc the only places we allow mutability currently are lazy initialisations of certain fields, but even in those cases the initialisation is deterministic and pure (aside from parsing exceptions). However in the `StateParty` case the initialisation must be done explicitly, it's non-deterministic (the mapping may change) *and* side-effecting. This is very error prone, and adds the mental burden of having to keep track of whether the key has been resolved or not.

Essentially the way `StateParty` is written now it's basically a cache for `PublicKey` -> `Party` lookups. Could we perhaps think of a better way to express this?

**Comment by rnicoll** [comment]
It doesn't use a sealed class or interface because the two shouldn't be interchangeable, the only way that might work is `Party` could be used in places that need `StateParty` and that leads to potential serialization problems (a `Party` could end up in a state and the whole thing serialized).

In terms of mutability, anything that needs the party should go through the resolution process.

Obviously a lot of further thought to be done, will come back to this Monday.

**Comment by mikehearn** [comment]
Yeah, I think the types involved here need a bit more thought. It'd have been better to put together a short design proposal before implementing this. I'm not at all convinced that introducing a new StateParty type is the way to go, although the notion of a party that can or cannot be resolved is worth representing in the type system. I was originally thinking something like this:

```
sealed class Party(val key: CompositeKey) {
    class ResolvedParty(key: CompositeKey, val identity: IdentityInfo) : Party
}

class IdentityInfo(val legalName: X500Name, val whatever ....)
```

where IdentityInfo would be fairly minimalist (more complex notions of identity can be handled locally by doing database joins on identity tables).

**Comment by exFalso** [comment]
Looking at the code in a bit more detail. If `StateParty` (or whatever we end up using) is indeed intended to be used in contract states and thus verification then this implies a major design change. In particular tx verification becomes context sensitive as it implicitly depends on the `PublicKey` -> `Party` mapping, which changes over time.

In essence previously `verify` was a `Transaction -> Bool` function, whereas with this change it becomes `(Transaction, NodeLocalContext) -> Bool`. Is this a change we want? It would make verification inconsistent across nodes!

Side note: types in the datamodel cannot have side-effecting member functions as per sandboxing rules

**Comment by mikehearn** [comment]
Inside the sandbox identity information is not meant to be available. Although, that said, a key is only meant to be owned by a single identity ever.

**Comment by rnicoll** [comment]
Pushing work in progress on rebuilding this with a sealed class, so you can see progress. I've also taken out the bit where the full party is ever inside the sandbox.

**Comment by exFalso** [comment]
This looks better! There are still some remains of `StateParty`, and `resolveParty` needs to be somewhere else (the sandbox won't load `Party.Anonymised` like this)

**Comment by rnicoll** [comment]
This should be ready for review, pending #181 being merged

File: node/src/main/kotlin/net/corda/node/utilities/JsonSupport.kt [hunk 1]
```diff
@@ -97,14 +99,14 @@ object JsonSupport {
 
     }
 
-    object PartySerializer : JsonSerializer<Party>() {
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/utilities/JsonSupport.kt]: Perhaps `PartyFullSerializer`

File: node/src/test/kotlin/net/corda/node/services/InMemoryNetworkMapCacheTest.kt [hunk 2]
```diff
@@ -24,19 +23,23 @@ class InMemoryNetworkMapCacheTest {
...
 
         // Node A currently knows only about itself, so this returns node A
-        assertEquals(nodeA.netMapCache.getNodeByLegalIdentityKey(keyPair.public.composite), nodeA.info)
+        assertEquals(nodeA.netMapCache.getNodeByLegalIdentityKey(nodeA.info.legalIdentity.owningKey), nodeA.info)
 
         databaseTransaction(nodeA.database) {
             nodeA.netMapCache.addNode(nodeB.info)
         }
         // Now both nodes match, so it throws an error
         expect<IllegalStateException> {
```
Comment (exFalso) [inline node/src/test/kotlin/net/corda/node/services/InMemoryNetworkMapCacheTest.kt]: Eh, this `expect` isn't doing anything! I think this should be some kind of `require` for exceptions

File: tools/explorer/src/main/kotlin/net/corda/explorer/Main.kt [hunk 3]
```diff
@@ -195,15 +195,15 @@ fun main(args: Array<String>) {
...
+                            issuerRPCGBP.deanonymiseParty(command.recipient)!!,
                             command.issueRef,
                             issuerNodeGBP.nodeInfo.legalIdentity)
                     Unit
                 }.generate(SplittableRandom())
                 issuerUSDEventGenerator.bankOfCordaIssueGenerator.map { command ->
                     issuerRPCUSD.startFlow(::IssuanceRequester,
                             command.amount,
-                            command.recipient,
+                            issuerRPCUSD.deanonymiseParty(command.recipient)!!,
```
Comment (exFalso) [inline tools/explorer/src/main/kotlin/net/corda/explorer/Main.kt]: `issuerRPCUSD`... this alongside `XMLHttpRequest` is the reason why I don't like all-capital abbreviations in camelcase. My tokeniser returns ambiguous results!

Comment (rnicoll) [inline tools/explorer/src/main/kotlin/net/corda/explorer/Main.kt]: I'll tackle that in a separate PR, to avoid mixing up cleanup vs actual changes.

Comment (exFalso) [inline tools/explorer/src/main/kotlin/net/corda/explorer/Main.kt]: Ah no worries I was just venting, we have lots of capital-in-camelcase symbols anyway...

**Comment by rnicoll** [comment]
@exFalso that should be your comments addressed

**Comment by mikehearn** [comment]
Please give me a bit of time to look at this, as it's a substantial change in an area where we didn't do much design work yet.

**Comment by rnicoll** [comment]
I think this one's grown too big to do anything with, it's trapped in rebase hell at the moment anyway. Suggest closing it, I'm pulling out the smaller parts and we can look at the bigger design in a later PR

**Pull request closed without merging**